### PR TITLE
HPCC-21997 Combine duplicate LDAP Security methods getResources

### PR DESCRIPTION
**Pull request by RussWhitehead**
Currently, in LDAP Security Manager, the 2 methods getResouces and getResourcesEx
are almost identical. This PR ccombines them into a single getResources call,
which handles the functionality of both. Note that an apparent bug is also
fixed that was previously in getResourcesEx, where the GetAttributesWrapper
looped through all attributes before saving, which meant only the last one was
saved. Also, pass the LDAP connection into the getManagedScopeTree method since
it is available

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [x] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by RussWhitehead** [comment]
This replaces previous https://github.com/hpcc-systems/HPCC-Platform/pull/12477

**Comment by RussWhitehead** [comment]
@richardkchapman Please merge

**Comment by richardkchapman** [comment]
@ghalliday I think this was something you requested?

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 1]
```diff
@@ -3461,19 +3461,26 @@ class CLdapClient : implements ILdapClient, public CInterface
...
+        return getResources(nullptr, rtype, basedn, prefix, searchstr, resources);
     }
 
-    virtual bool getResources(LDAP* ld, SecResourceType rtype, const char * basedn, const char* prefix, IArrayOf<ISecResource>& resources)
+    virtual bool getResources(LDAP* ld, SecResourceType rtype, const char * basedn, const char * prefix, const char * searchstr, IArrayOf<ISecResource>& resources)
     {
+        Owned<ILdapConnection> lconn;
+        if (nullptr == ld)
+        {
+            lconn.set(m_connections->getConnection());
```
Comment (ghalliday) [inline system/security/LdapSecurity/ldapconnection.cpp]: This leaks - should be setown.

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 2]
```diff
@@ -6180,15 +6114,20 @@ class CLdapClient : implements ILdapClient, public CInterface
         return addResource(RT_FILE_SCOPE, user, resource, PT_ADMINISTRATORS_AND_USER, m_ldapconfig->getResourceBasedn(RT_FILE_SCOPE));
     }
 
-    virtual aindex_t getManagedScopeTree(SecResourceType rtype, const char * basedn, IArrayOf<ISecResource>& scopes)
+    virtual aindex_t getManagedScopeTree(LDAP* ld, SecResourceType rtype, const char * basedn, IArrayOf<ISecResource>& scopes)
     {
+        Owned<ILdapConnection> lconn;
+        if (nullptr == ld)
+        {
+            lconn.set(m_connections->getConnection());
```
Comment (ghalliday) [inline system/security/LdapSecurity/ldapconnection.cpp]: This also leaks - should be setown()

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 3]
```diff
@@ -6180,15 +6114,20 @@ class CLdapClient : implements ILdapClient, public CInterface
...
+            ld = ((CLdapConnection*)lconn.get())->getLd();
```
Comment (ghalliday) [inline system/security/LdapSecurity/ldapconnection.cpp]: Why is this cast here? This function being called is the only function in the interface.  Same for the other code above.

**Review by ghalliday — COMMENTED** [review]
@RussWhitehead a few minor issues to clean up.

**Comment by ghalliday** [comment]
@RussWhitehead looks good.  Please squash and it can be merged.

**Comment by RussWhitehead** [comment]
@ghalliday Squashed, please merge

**Pull request merged**